Bitcoin Core refactoring and cleanups 1

str4d opened the pull request:
Cherry-picked from the following upstream PRs:

- bitcoin/bitcoin#6538
- bitcoin/bitcoin#6163
- bitcoin/bitcoin#6982
- bitcoin/bitcoin#6986
- bitcoin/bitcoin#7053
- bitcoin/bitcoin#7444
- bitcoin/bitcoin#7793
  - Excluding some comments in `txmempool.h` on code we haven't yet pulled in.
- bitcoin/bitcoin#7916
- bitcoin/bitcoin#7815

Additionally, the Equihash parameters are moved into the consensus parameters, and
various other functions in `main.cpp` have `const CChainParams&` arguments added.

Part of #2074.

str4d commented:
Came across these as I was working on #2905, but it turned out that the merge conflict this resolved was very minor relative to the size of this PR, so I'm keeping it separate.

str4d commented:
Rebased on master to fix merge conflicts. I also pulled in bitcoin/bitcoin#6986 which I'd previously missed, which required pulling in bitcoin/bitcoin#5994 (taking the commits I'd already cherry-picked in #2225).

str4d commented:
@zkbot try

str4d commented:
Most of the Zcash-specific RPC tests failed because they assume that the first `getnewaddress` call will return the address the miner is using, but after bitcoin/bitcoin#5994 that no longer appears to be the case. I'm trying to figure out what exactly caused the change in behaviour, but the end result [appears to be intentional](https://github.com/bitcoin/bitcoin/pull/5994#discussion-diff-29037319), so I'll adjust the RPC tests to grab the addresses from the mined coinbase instead.

str4d commented:
Missed a more complex coinbase case in `wallet_shieldcoinbase`, and refactored my commit to handle it better.

@zkbot try

str4d commented:
One last bugfix (due to this PR, the cached chain used by many RPC tests uses a single address per coinbase).

@zkbot try

str4d commented:
`debian9` builder failed due to a transient RPC test failure. However, the `debian8-kitchensink` builder uncovered a build error due to not updating the AMQP code (which is not enabled by default on other builders).

str4d commented:
Force-pushed to update the `ReadBlockFromDisk` calls in the AMQP code. This PR should be ready to go now.

str4d commented:
Rebased on master to fix merge conflicts.

str4d commented:
Rebased on master to fix merge conflict.

str4d commented:
Rebased on master to fix merge conflict.

str4d commented:
@zkbot try

bitcartel commented:
The `wallet_anchorfork` (cache invalidation bug) test failed.

str4d commented:
Looking into the test failure above today.

str4d commented:
Aha, simple bug (missed that test while updating others to use `get_coinbase_address()`).

str4d commented:
Force-pushed to edit the fix into the final commit.

@zkbot try

str4d commented:
Rebased on master to fix merge conflicts.

@zkbot try

str4d commented:
Rebased on master to fix merge conflicts.

@zkbot try

str4d commented:
Force-pushed to update more recent RPC tests for the coinbase address changes.

@zkbot try

mms710 commented:
@str4d is it possible for you to split this up into a few smaller PRs? We can get them reviewed more easily if they're smaller.

str4d commented:
@mms710 sure; I'll rebase this PR to fix the merge conflicts, and then I'll create a new PR with the bitcoin/bitcoin#5994 changes. I can't easily split this PR into multiple parallel PRs because IIRC this set of PRs all build on each other, so they need to be merged sequentially. But I can break it up into one new PR at a time containing a subset of these changes, making the next PR after the previous one gets merged.

str4d commented:
@mms710 I've opened #3762 with the first 5 commits from this PR.

mms710 commented:
@str4d thanks so much! I'll make sure it gets added near the top of the PRs to be reviewed queue.

str4d commented:
Rebased on master to remove the commits merged in #3762. I also added more commits to remove all Zcash-introduced usages of the `Params()` global inside functions that had no such usages upstream.

str4d commented:
@mms710 Now that the #3762 commits have been removed, this PR should be easily reviewable commit-by-commit. All remaining commits are short, trivial refactors.

mms710 commented:
@str4d sweet! I just put it at the top of the PR backlog

Eirik0 commented on src/pow.cpp:
```diff
@@ -91,10 +91,10 @@ unsigned int CalculateNextWorkRequired(arith_uint256 bnAvg,
     return bnNew.GetCompact();
 }
 
-bool CheckEquihashSolution(const CBlockHeader *pblock, const CChainParams& params)
```
Nit: there are a few places in this commit where we just say `params` rather than saying `chainparams` or `consensusParams` explicitly.

str4d replied:
IIRC these were places that we ourselves introduced parameters in the past. I didn't alter the variable names to make the PR smaller, but I can push another commit that makes them consistent.

str4d replied:
I decided to only address this in `main.cpp`; after this PR only `Consensus::Params` is used in `pow.cpp`, so there's not much point in adding additional change lines to this PR. We can address later if desired.

str4d commented:
Rebased on master to fix a merge conflict.

str4d commented:
Addressed @Eirik0's comment for `main.cpp`, and also added another commit to address `IsInitialBlockDownload()` which I had missed previously.

str4d commented:
Rebased on master to address merge conflicts.

@zkbot try

str4d commented:
Force-pushed to fix a bug introduced into a test during merge conflict resolution (the parameters were being cached across an upgrade activation, when they should have been fetched anew).

@zkbot try

str4d commented:
Test failure above was transient AFAICT.

str4d commented:
Rebased on master to fix merge conflicts.

Eirik0 commented:
@zkbot try

str4d commented:
Kitchen sink builder failed with `--enable-proton`, but that works for me locally, so I'm hoping this was a transient failure.

@zkbot retry

str4d commented on src/init.cpp:
```diff
@@ -1868,7 +1868,7 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 #ifdef ENABLE_MINING
     // Generate coins in the background
-    GenerateBitcoins(GetBoolArg("-gen", false), GetArg("-genproclimit", 1), Params());
+    GenerateBitcoins(GetBoolArg("-gen", false), GetArg("-genproclimit", 1), chainparams);
```
Mmm. There are still a bunch more places that need refactoring in this way, so in the interest of PR size, let's leave that to `2074-main-refactor-2` 😄

str4d commented:
I see above that @Eirik0 thumbed-up the changes I made after his review, and since then the only changes to this PR have been fixing merge conflicts. Any further comments can be addressed in subsequent PRs.

@zkbot r+
